### PR TITLE
Refactor maze generator

### PR DESCRIPTION
**Pull request by jensj12**
Split class into 3 files, use and object of the class and use directions for deciding the next point

### CODE REVIEW

File: MeesGame/Gameplay/Generator/MazeGeneratorInit.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                    loopChance = 0.01;
+                    loopDoorChance = 0.3;
+                    loopHoleChance = 0.7;
+                    splitChance = 0.02;
+                    doorChance = 0.08;
+                    numKeys = 1;
+                    break;
+            }
+
+            // Make sure the dimensions of the maze are add and reasonable.
```
Comment (LisanneW) [inline MeesGame/Gameplay/Generator/MazeGeneratorInit.cs]: "odd and reasonable" I presume

File: MeesGame/Gameplay/Generator/MazeGeneratorMain.cs [hunk 2]
```diff
@@ -0,0 +1,148 @@
+﻿using MeesGame;
+using Microsoft.Xna.Framework;
+using System;
+using System.Collections.Generic;
```
Comment (LisanneW) [inline MeesGame/Gameplay/Generator/MazeGeneratorMain.cs]: Unnecessary usings.

File: MeesGame/Gameplay/Generator/MazeGeneratorMain.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            }
+
+            // Add the startTile.
+            tiles.Add(new StartTile(0, "playerstart"), start.X, start.Y);
+
+            // Add an end tile on the best exit.
+            tiles.Add(new EndTile(), bestExit.X, bestExit.Y);
+        }
+    }
+}
```
Comment (LisanneW) [inline MeesGame/Gameplay/Generator/MazeGeneratorMain.cs]: Missing newline.

File: MeesGame/Gameplay/Generator/MazeGeneratorInit.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        private Point bestExit;
+        private int[,] exitScore;
+        private IList<Point> nodesToDo;
+        private int totalNodesToDo;
+        private int[] keyNodes;
+        private Point start;
+
+        private void InitMazeGen(int numRows, int numCols, int difficulty)
+        {
+            int numKeys = 0;
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/Generator/MazeGeneratorInit.cs]: This should be a member variable like the others

File: MeesGame/Gameplay/Generator/MazeGeneratorInit.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        private double doorChance = 0;
+        private TileField tiles;
+        private Point bestExit;
+        private int[,] exitScore;
+        private IList<Point> nodesToDo;
+        private int totalNodesToDo;
+        private int[] keyNodes;
+        private Point start;
+
+        private void InitMazeGen(int numRows, int numCols, int difficulty)
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/Generator/MazeGeneratorInit.cs:33]: You could split this into some helper methods.

```c#
useDifficulty(difficulty);
makeSureDimensionsAreValid();
createTileFieldWithJustWalls();
initNodesToDo();
initKeyNodes();
initExitScores();
```

File: MeesGame/Gameplay/Generator/MazeGeneratorMain.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            bool randomNext = false; // Is set to true if the current path can't be extended
+            Tile tileToAdd;
+            int nodesDone = -1; // Makes it 0 in the first iteration.
+
+            // Add hallways until all tiles are filled (main algorithm).
+            while (nodesToDo.Count != 0)
+            {
+                nodesDone += 1;
+
+                // Continue from the last node, or take a random one
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/Generator/MazeGeneratorMain.cs:30]: You could make this a helper function:

    Point current = determineNextNode();

And then later you can do

    nodesToDo.RemoveAt(nodesToDo.count-1);

File: MeesGame/Gameplay/Generator/MazeGeneratorMain.cs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                {
+                    // Swap the last node with another node
+                    int nodeToDoNext = random.Next(nodesToDo.Count);
+                    swap(nodesToDo, nodeToDoNext, position);
+                }
+
+                Point current = nodesToDo[position];
+                Point next;
+
+                // Make an arrray and fill it with 1-4 in a random order.
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/Generator/MazeGeneratorMain.cs]: Comment does not match function name

File: MeesGame/Gameplay/Generator/MazeGeneratorMain.cs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+                                // Or add a door instead of making it a loop, you don't need to check if a key has been placed for this door because it's placed in a loop.
+                                if (random.Next(BIGINT) < loopDoorChance * BIGINT)
+                                {
+                                    tileToAdd = new DoorTile();
+                                }
+
+                                // Add the tile.
+                                int randNumb = random.Next(BIGINT);
+                                if (randNumb < (loopDoorChance + loopHoleChance) * BIGINT) tileToAdd = new DoorTile();
+                                if (randNumb < loopHoleChance * BIGINT) tileToAdd = new HoleTile();
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/Generator/MazeGeneratorMain.cs]: I found it hard to see what was going on at a first glance. Do you think this is clearer?

```c#
if (randNumb < loopHoleChance * BIGINT) tileToAdd = new HoleTile();
else if (randNumb < (loopDoorChance + loopHoleChance) * BIGINT) tileToAdd = new DoorTile();
else tileToAdd = new FloorTile();

**Review by maxmeijer42 — CHANGES_REQUESTED** [review]

File: MeesGame/Gameplay/Generator/MazeGeneratorMain.cs (hunk 6 above)
Comment (jensj12) [reply MeesGame/Gameplay/Generator/MazeGeneratorMain.cs:30]: Using ```nodesToDo.RemoveAt(nodesToDo.count-1);``` At the end will cause newly added tiles to be removed from the list.

File: MeesGame/Gameplay/Generator/MazeGeneratorMain.cs (hunk 8 above)
Comment (jensj12) [reply MeesGame/Gameplay/Generator/MazeGeneratorMain.cs]: I've done something similar in a helper function, and fixed a calculation bug with it.

**Review by maxmeijer42 — APPROVED** [review]
Nice, code is much cleaner now :)

**Pull request merged**